2.3.3 staged: native-image build fails for helidon-quickstart-se

I'm testing the staged bits for 2.3.3 and hit this issue. Note that it does NOT seem to be an issue on the 2.4.0 (master) branch. So maybe we missed something in a backport.

JDK: Java 11.0.12
GraalVM: graalvm-ce-java11-21.1.0 and graalvm-ce-java11-21.0.0.2

Building native image for `examples/quickstarts/helidon-quickstart-se` (and likely others) fails. See errors below. To reproduce:

```
git checkout tags/2.3.3
mvn clean install -DskipTests
cd examples/quickstarts/helidon-quickstart-se
mvn clean install -Pnative-image 
```
Errors:

```
[INFO] --- helidon-maven-plugin:2.2.3:native-image (native-image) @ helidon-quickstart-se ---
[INFO] Building native image :/Users/jdipol/GitHub/barchetta/helidon/examples/quickstarts/helidon-quickstart-se/target/helidon-quickstart-se
[INFO] [helidon-quickstart-se:78761]    classlist:   1,548.15 ms,  0.96 GB
[INFO] [helidon-quickstart-se:78761]        (cap):   2,039.95 ms,  0.96 GB
[INFO] [helidon-quickstart-se:78761]        setup:   3,934.08 ms,  0.96 GB
[INFO] 2021.09.04 13:00:02 INFO io.helidon.common.LogConfig Thread[ForkJoinPool-4-worker-3,5,main]: Logging at initialization configured using classpath: /logging.properties
[INFO] [helidon-quickstart-se:78761]     (clinit):     622.19 ms,  3.24 GB
[INFO] [helidon-quickstart-se:78761]   (typeflow):  14,477.40 ms,  3.24 GB
[INFO] [helidon-quickstart-se:78761]    (objects):  14,591.91 ms,  3.24 GB
[INFO] [helidon-quickstart-se:78761]   (features):   1,258.22 ms,  3.24 GB
[INFO] [helidon-quickstart-se:78761]     analysis:  32,295.08 ms,  3.24 GB
[WARNING] Error: Unsupported features in 3 methods
[WARNING] Detailed message:
[WARNING] Error: Detected a started Thread in the image heap. Threads running in the image generator are no longer running at image runtime.  To see how this object got instantiated use --trace-object-instantiation=java.lang.Thread. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
[WARNING] Trace: Object was reached by
[WARNING] 	reading field java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.leader of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue@314c3331 reached by
[WARNING] 	reading field java.util.concurrent.ThreadPoolExecutor.workQueue of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor@6051e655 reached by
[WARNING] 	reading field java.util.concurrent.Executors$DelegatedExecutorService.e of
[WARNING] 		constant java.util.concurrent.Executors$DelegatedScheduledExecutorService@2fa9e950 reached by
[WARNING] 	scanning method io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] Call path from entry point to io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown():
[WARNING] 	at io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] 	at io.helidon.metrics.MetricsSupport.onShutdown(MetricsSupport.java:435)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport.webServerStopped(HelidonRestServiceSupport.java:114)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport$$Lambda$1198/0x00000007c10f6c40.run(Unknown Source)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeHooks(RuntimeSupport.java:125)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeStartupHooks(RuntimeSupport.java:75)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:137)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:182)
[WARNING] 	at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)
[WARNING] Error: Detected a started Thread in the image heap. Threads running in the image generator are no longer running at image runtime.  To see how this object got instantiated use --trace-object-instantiation=java.lang.Thread. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
[WARNING] Trace: Object was reached by
[WARNING] 	reading field java.util.concurrent.ThreadPoolExecutor$Worker.thread of
[WARNING] 		constant java.util.concurrent.ThreadPoolExecutor$Worker@2cb46cfb reached by
[WARNING] 	reading field java.util.HashMap$Node.key of
[WARNING] 		constant java.util.HashMap$Node@e8a03e1 reached by
[WARNING] 	indexing into array
[WARNING] 		constant java.util.HashMap$Node[]@4936b3d0 reached by
[WARNING] 	reading field java.util.HashMap.table of
[WARNING] 		constant java.util.HashMap@4ce83e22 reached by
[WARNING] 	reading field java.util.HashSet.map of
[WARNING] 		constant java.util.HashSet@7dc67139 reached by
[WARNING] 	reading field java.util.concurrent.ThreadPoolExecutor.workers of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor@6051e655 reached by
[WARNING] 	reading field java.util.concurrent.Executors$DelegatedExecutorService.e of
[WARNING] 		constant java.util.concurrent.Executors$DelegatedScheduledExecutorService@2fa9e950 reached by
[WARNING] 	scanning method io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] Call path from entry point to io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown():
[WARNING] 	at io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] 	at io.helidon.metrics.MetricsSupport.onShutdown(MetricsSupport.java:435)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport.webServerStopped(HelidonRestServiceSupport.java:114)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport$$Lambda$1198/0x00000007c10f6c40.run(Unknown Source)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeHooks(RuntimeSupport.java:125)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeStartupHooks(RuntimeSupport.java:75)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:137)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:182)
[WARNING] 	at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)
[WARNING] Error: Detected a started Thread in the image heap. Threads running in the image generator are no longer running at image runtime.  To see how this object got instantiated use --trace-object-instantiation=java.lang.Thread. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
[WARNING] Trace: Object was reached by
[WARNING] 	reading field java.util.concurrent.locks.AbstractQueuedSynchronizer$Node.thread of
[WARNING] 		constant java.util.concurrent.locks.AbstractQueuedSynchronizer$Node@11e8bafe reached by
[WARNING] 	reading field java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.firstWaiter of
[WARNING] 		constant java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@3af3d3f0 reached by
[WARNING] 	reading field java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.available of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue@314c3331 reached by
[WARNING] 	reading field java.util.concurrent.ThreadPoolExecutor.workQueue of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor@6051e655 reached by
[WARNING] 	reading field java.util.concurrent.Executors$DelegatedExecutorService.e of
[WARNING] 		constant java.util.concurrent.Executors$DelegatedScheduledExecutorService@2fa9e950 reached by
[WARNING] 	scanning method io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] Call path from entry point to io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown():
[WARNING] 	at io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] 	at io.helidon.metrics.MetricsSupport.onShutdown(MetricsSupport.java:435)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport.webServerStopped(HelidonRestServiceSupport.java:114)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport$$Lambda$1198/0x00000007c10f6c40.run(Unknown Source)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeHooks(RuntimeSupport.java:125)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeStartupHooks(RuntimeSupport.java:75)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:137)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:182)
[WARNING] 	at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)
[WARNING]
[WARNING] com.oracle.svm.core.util.UserError$UserException: Unsupported features in 3 methods
[WARNING] Detailed message:
[WARNING] Error: Detected a started Thread in the image heap. Threads running in the image generator are no longer running at image runtime.  To see how this object got instantiated use --trace-object-instantiation=java.lang.Thread. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
[WARNING] Trace: Object was reached by
[WARNING] 	reading field java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.leader of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue@314c3331 reached by
[WARNING] 	reading field java.util.concurrent.ThreadPoolExecutor.workQueue of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor@6051e655 reached by
[WARNING] 	reading field java.util.concurrent.Executors$DelegatedExecutorService.e of
[WARNING] 		constant java.util.concurrent.Executors$DelegatedScheduledExecutorService@2fa9e950 reached by
[WARNING] 	scanning method io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] Call path from entry point to io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown():
[WARNING] 	at io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] 	at io.helidon.metrics.MetricsSupport.onShutdown(MetricsSupport.java:435)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport.webServerStopped(HelidonRestServiceSupport.java:114)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport$$Lambda$1198/0x00000007c10f6c40.run(Unknown Source)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeHooks(RuntimeSupport.java:125)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeStartupHooks(RuntimeSupport.java:75)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:137)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:182)
[WARNING] 	at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)
[WARNING] Error: Detected a started Thread in the image heap. Threads running in the image generator are no longer running at image runtime.  To see how this object got instantiated use --trace-object-instantiation=java.lang.Thread. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
[WARNING] Trace: Object was reached by
[WARNING] 	reading field java.util.concurrent.ThreadPoolExecutor$Worker.thread of
[WARNING] 		constant java.util.concurrent.ThreadPoolExecutor$Worker@2cb46cfb reached by
[WARNING] 	reading field java.util.HashMap$Node.key of
[WARNING] 		constant java.util.HashMap$Node@e8a03e1 reached by
[WARNING] 	indexing into array
[WARNING] 		constant java.util.HashMap$Node[]@4936b3d0 reached by
[WARNING] 	reading field java.util.HashMap.table of
[WARNING] 		constant java.util.HashMap@4ce83e22 reached by
[WARNING] 	reading field java.util.HashSet.map of
[WARNING] 		constant java.util.HashSet@7dc67139 reached by
[WARNING] 	reading field java.util.concurrent.ThreadPoolExecutor.workers of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor@6051e655 reached by
[WARNING] 	reading field java.util.concurrent.Executors$DelegatedExecutorService.e of
[WARNING] 		constant java.util.concurrent.Executors$DelegatedScheduledExecutorService@2fa9e950 reached by
[WARNING] 	scanning method io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] Call path from entry point to io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown():
[WARNING] 	at io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] 	at io.helidon.metrics.MetricsSupport.onShutdown(MetricsSupport.java:435)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport.webServerStopped(HelidonRestServiceSupport.java:114)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport$$Lambda$1198/0x00000007c10f6c40.run(Unknown Source)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeHooks(RuntimeSupport.java:125)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeStartupHooks(RuntimeSupport.java:75)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:137)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:182)
[WARNING] 	at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)
[WARNING] Error: Detected a started Thread in the image heap. Threads running in the image generator are no longer running at image runtime.  To see how this object got instantiated use --trace-object-instantiation=java.lang.Thread. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
[WARNING] Trace: Object was reached by
[WARNING] 	reading field java.util.concurrent.locks.AbstractQueuedSynchronizer$Node.thread of
[WARNING] 		constant java.util.concurrent.locks.AbstractQueuedSynchronizer$Node@11e8bafe reached by
[WARNING] 	reading field java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.firstWaiter of
[WARNING] 		constant java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@3af3d3f0 reached by
[WARNING] 	reading field java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.available of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue@314c3331 reached by
[WARNING] 	reading field java.util.concurrent.ThreadPoolExecutor.workQueue of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor@6051e655 reached by
[WARNING] 	reading field java.util.concurrent.Executors$DelegatedExecutorService.e of
[WARNING] 		constant java.util.concurrent.Executors$DelegatedScheduledExecutorService@2fa9e950 reached by
[WARNING] 	scanning method io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] Call path from entry point to io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown():
[WARNING] 	at io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] 	at io.helidon.metrics.MetricsSupport.onShutdown(MetricsSupport.java:435)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport.webServerStopped(HelidonRestServiceSupport.java:114)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport$$Lambda$1198/0x00000007c10f6c40.run(Unknown Source)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeHooks(RuntimeSupport.java:125)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeStartupHooks(RuntimeSupport.java:75)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:137)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:182)
[WARNING] 	at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)
[WARNING]
[WARNING] 	at com.oracle.svm.core.util.UserError.abort(UserError.java:82)
[WARNING] 	at com.oracle.svm.hosted.FallbackFeature.reportAsFallback(FallbackFeature.java:233)
[WARNING] 	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:798)
[WARNING] 	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:582)
[WARNING] 	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$2(NativeImageGenerator.java:495)
[WARNING] 	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
[WARNING] 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
[WARNING] 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
[WARNING] 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
[WARNING] 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
[WARNING] 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
[WARNING] Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Unsupported features in 3 methods
[WARNING] Detailed message:
[WARNING] Error: Detected a started Thread in the image heap. Threads running in the image generator are no longer running at image runtime.  To see how this object got instantiated use --trace-object-instantiation=java.lang.Thread. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
[WARNING] Trace: Object was reached by
[WARNING] 	reading field java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.leader of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue@314c3331 reached by
[WARNING] 	reading field java.util.concurrent.ThreadPoolExecutor.workQueue of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor@6051e655 reached by
[WARNING] 	reading field java.util.concurrent.Executors$DelegatedExecutorService.e of
[WARNING] 		constant java.util.concurrent.Executors$DelegatedScheduledExecutorService@2fa9e950 reached by
[WARNING] 	scanning method io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] Call path from entry point to io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown():
[WARNING] 	at io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] 	at io.helidon.metrics.MetricsSupport.onShutdown(MetricsSupport.java:435)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport.webServerStopped(HelidonRestServiceSupport.java:114)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport$$Lambda$1198/0x00000007c10f6c40.run(Unknown Source)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeHooks(RuntimeSupport.java:125)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeStartupHooks(RuntimeSupport.java:75)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:137)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:182)
[WARNING] 	at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)
[WARNING] Error: Detected a started Thread in the image heap. Threads running in the image generator are no longer running at image runtime.  To see how this object got instantiated use --trace-object-instantiation=java.lang.Thread. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
[WARNING] Trace: Object was reached by
[WARNING] 	reading field java.util.concurrent.ThreadPoolExecutor$Worker.thread of
[WARNING] 		constant java.util.concurrent.ThreadPoolExecutor$Worker@2cb46cfb reached by
[WARNING] 	reading field java.util.HashMap$Node.key of
[WARNING] 		constant java.util.HashMap$Node@e8a03e1 reached by
[WARNING] 	indexing into array
[WARNING] 		constant java.util.HashMap$Node[]@4936b3d0 reached by
[WARNING] 	reading field java.util.HashMap.table of
[WARNING] 		constant java.util.HashMap@4ce83e22 reached by
[WARNING] 	reading field java.util.HashSet.map of
[WARNING] 		constant java.util.HashSet@7dc67139 reached by
[WARNING] 	reading field java.util.concurrent.ThreadPoolExecutor.workers of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor@6051e655 reached by
[WARNING] 	reading field java.util.concurrent.Executors$DelegatedExecutorService.e of
[WARNING] 		constant java.util.concurrent.Executors$DelegatedScheduledExecutorService@2fa9e950 reached by
[WARNING] 	scanning method io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] Call path from entry point to io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown():
[WARNING] 	at io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] 	at io.helidon.metrics.MetricsSupport.onShutdown(MetricsSupport.java:435)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport.webServerStopped(HelidonRestServiceSupport.java:114)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport$$Lambda$1198/0x00000007c10f6c40.run(Unknown Source)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeHooks(RuntimeSupport.java:125)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeStartupHooks(RuntimeSupport.java:75)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:137)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:182)
[WARNING] 	at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)
[WARNING] Error: Detected a started Thread in the image heap. Threads running in the image generator are no longer running at image runtime.  To see how this object got instantiated use --trace-object-instantiation=java.lang.Thread. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
[WARNING] Trace: Object was reached by
[WARNING] 	reading field java.util.concurrent.locks.AbstractQueuedSynchronizer$Node.thread of
[WARNING] 		constant java.util.concurrent.locks.AbstractQueuedSynchronizer$Node@11e8bafe reached by
[WARNING] 	reading field java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.firstWaiter of
[WARNING] 		constant java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@3af3d3f0 reached by
[WARNING] 	reading field java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue.available of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor$DelayedWorkQueue@314c3331 reached by
[WARNING] 	reading field java.util.concurrent.ThreadPoolExecutor.workQueue of
[WARNING] 		constant java.util.concurrent.ScheduledThreadPoolExecutor@6051e655 reached by
[WARNING] 	reading field java.util.concurrent.Executors$DelegatedExecutorService.e of
[WARNING] 		constant java.util.concurrent.Executors$DelegatedScheduledExecutorService@2fa9e950 reached by
[WARNING] 	scanning method io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] Call path from entry point to io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown():
[WARNING] 	at io.helidon.metrics.ExponentiallyDecayingReservoir.onServerShutdown(ExponentiallyDecayingReservoir.java:79)
[WARNING] 	at io.helidon.metrics.MetricsSupport.onShutdown(MetricsSupport.java:435)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport.webServerStopped(HelidonRestServiceSupport.java:114)
[WARNING] 	at io.helidon.servicecommon.rest.HelidonRestServiceSupport$$Lambda$1198/0x00000007c10f6c40.run(Unknown Source)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeHooks(RuntimeSupport.java:125)
[WARNING] 	at com.oracle.svm.core.jdk.RuntimeSupport.executeStartupHooks(RuntimeSupport.java:75)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:137)
[WARNING] 	at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:182)
[WARNING] 	at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)
[WARNING]
[WARNING] 	at com.oracle.graal.pointsto.constraints.UnsupportedFeatures.report(UnsupportedFeatures.java:129)
[WARNING] 	at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:795)
[WARNING] 	... 8 more
[WARNING] Error: Image build request failed with exit status 1
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  57.821 s
[INFO] Finished at: 2021-09-04T13:00:35-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.helidon.build-tools:helidon-maven-plugin:2.2.3:native-image (native-image) on project helidon-quickstart-se: Image generation failed, exit code: 1 -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

--- a/examples/quickstarts/helidon-quickstart-se/src/main/resources/logging.properties
+++ b/examples/quickstarts/helidon-quickstart-se/src/main/resources/logging.properties
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2018, 2019 Oracle and/or its affiliates. All rights reserved.
+# Copyright (c) 2018, 2021 Oracle and/or its affiliates.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
 # you may not use this file except in compliance with the License.

--- a/examples/quickstarts/helidon-quickstart-se/src/main/resources/logging.properties
+++ b/examples/quickstarts/helidon-quickstart-se/src/main/resources/logging.properties
@@ -32,3 +32,4 @@ java.util.logging.SimpleFormatter.format=%1$tY.%1$tm.%1$td %1$tH:%1$tM:%1$tS %4$
 #io.helidon.security.level=INFO
 #io.helidon.common.level=INFO
 #io.netty.level=INFO
+io.helidon.metrics.ExponentiallyDecayingReservoir.level=FINER

--- a/examples/quickstarts/helidon-quickstart-se/src/main/resources/logging.properties
+++ b/examples/quickstarts/helidon-quickstart-se/src/main/resources/logging.properties
@@ -32,4 +32,3 @@ java.util.logging.SimpleFormatter.format=%1$tY.%1$tm.%1$td %1$tH:%1$tM:%1$tS %4$
 #io.helidon.security.level=INFO
 #io.helidon.common.level=INFO
 #io.netty.level=INFO
-io.helidon.metrics.ExponentiallyDecayingReservoir.level=FINER

--- a/metrics/metrics/pom.xml
+++ b/metrics/metrics/pom.xml
@@ -102,6 +102,7 @@
                             <configuration>
                                 <excludes>
                                     <exclude>**/TestServerWithKeyPerformanceIndicatorMetrics.java</exclude>
+                                    <exclude>**/TestExponentiallyDecayingReservoir.java</exclude>
                                 </excludes>
                             </configuration>
                         </execution>
@@ -129,6 +130,19 @@
                             <configuration>
                                 <includes>
                                     <include>**/TestServerWithKeyPerformanceIndicatorMetrics.java</include>
+                                </includes>
+                            </configuration>
+                        </execution>
+                        <execution>
+                            <!-- Run in separate invocation so we have only one task updating
+                                 ExponentiallyDecayingReservoir current time values. -->
+                            <id>exponentially-decaying-reservoir-check</id>
+                            <goals>
+                                <goal>test</goal>
+                            </goals>
+                            <configuration>
+                                <includes>
+                                    <include>**/TestExponentiallyDecayingReservoir.java</include>
                                 </includes>
                             </configuration>
                         </execution>

--- a/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
@@ -16,16 +16,13 @@
 
 package io.helidon.metrics;
 
+import java.time.Duration;
 import java.util.ArrayList;
-import java.util.List;
 import java.util.concurrent.ConcurrentSkipListMap;
-import java.util.concurrent.Executors;
-import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ThreadLocalRandom;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
-import java.util.logging.Level;
 import java.util.logging.Logger;
 
 import static java.lang.Math.exp;
@@ -55,27 +52,14 @@ import static java.lang.Math.min;
  * registers its own {@code Runnable} which updates its own value, and the single executor invokes all of them when it runs.
  */
 class ExponentiallyDecayingReservoir {
+
     private static final int DEFAULT_SIZE = 1028;
     private static final double DEFAULT_ALPHA = 0.015;
     private static final long RESCALE_THRESHOLD = TimeUnit.HOURS.toNanos(1);
 
-    /*
-     * Avoid computing the current time in seconds during every reservoir update by updating its value on a scheduled basis.
-     */
-    private static final long CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS = 250;
-
-    private static final List<ExponentiallyDecayingReservoir> RESERVOIRS = new ArrayList<>();
-
     private static final Logger LOGGER = Logger.getLogger(ExponentiallyDecayingReservoir.class.getName());
-    private static volatile ScheduledExecutorService currentTimeUpdaterExecutorService;
 
-    private enum CurrentTimeExecutorServiceState {
-        DORMANT, // never started
-        STARTED, // started and still running
-        STOPPED  // stopped
-        }
-
-    private static CurrentTimeExecutorServiceState currentTimeExecutorServiceState = CurrentTimeExecutorServiceState.DORMANT;
+    private static final Duration CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL = Duration.ofMillis(250);
 
     private final ConcurrentSkipListMap<Double, WeightedSnapshot.WeightedSample> values;
     private final ReentrantReadWriteLock lock;
@@ -108,76 +92,10 @@ class ExponentiallyDecayingReservoir {
         this.alpha = alpha;
         this.size = size;
         this.count = new AtomicLong(0);
-        RESERVOIRS.add(this);
-        updateCurrentTimeInSeconds();
+        PeriodicExecutor.enroll(this::updateTimeInSeconds, CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL);
+        this.updateTimeInSeconds();
         this.startTime = currentTimeInSeconds;
         this.nextScaleTime = new AtomicLong(clock.nanoTick() + RESCALE_THRESHOLD);
-    }
-
-    static synchronized void init() {
-        if (currentTimeExecutorServiceState == CurrentTimeExecutorServiceState.DORMANT) {
-            currentTimeExecutorServiceState = CurrentTimeExecutorServiceState.STARTED;
-            currentTimeUpdaterExecutorService = startCurrentTimeExecutor();
-        }
-    }
-
-    // we only expect a single instance of MetricSupport registered
-    // with webserver, so we can stop the executor when the associated webserver stops
-    static synchronized void onServerShutdown() {
-        if (currentTimeExecutorServiceState == CurrentTimeExecutorServiceState.STARTED) {
-            stopCurrentTimeExecutor();
-        } else {
-            LOGGER.log(Level.WARNING, String.format(
-                    "Attempt to start current time executor; expected state %s but found %s; ignored",
-                    CurrentTimeExecutorServiceState.DORMANT,
-                    currentTimeExecutorServiceState),
-                new IllegalStateException());
-        }
-        currentTimeExecutorServiceState = CurrentTimeExecutorServiceState.STOPPED;
-    }
-
-    private static ScheduledExecutorService startCurrentTimeExecutor() {
-        LOGGER.log(Level.FINE, () -> String.format("Starting %s current time executor with interval %d ms",
-                ExponentiallyDecayingReservoir.class.getSimpleName(), CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS));
-        ScheduledExecutorService result = Executors.newSingleThreadScheduledExecutor();
-        result.scheduleAtFixedRate(ExponentiallyDecayingReservoir::updateCurrentTimeInSecondsForAllReservoirs,
-                CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS,
-                CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS,
-                TimeUnit.MILLISECONDS);
-        return result;
-    }
-
-    private static void stopCurrentTimeExecutor() {
-        switch (currentTimeExecutorServiceState) {
-            case STARTED:
-                shutdownCurrentTimeExecutor();
-                break;
-
-            case STOPPED:
-            case DORMANT:
-                break;
-
-            default:
-                LOGGER.log(Level.WARNING, "Attempt to stop current time executor found unexpected state "
-                        + currentTimeExecutorServiceState);
-                shutdownCurrentTimeExecutor();
-        }
-    }
-
-    private static void shutdownCurrentTimeExecutor() {
-        LOGGER.log(Level.FINE, String.format("Shutting down %s current time executor",
-                ExponentiallyDecayingReservoir.class.getSimpleName()));
-        currentTimeUpdaterExecutorService.shutdownNow();
-    }
-
-    private static void updateCurrentTimeInSecondsForAllReservoirs() {
-        for (ExponentiallyDecayingReservoir reservoir : RESERVOIRS) {
-            reservoir.updateCurrentTimeInSeconds();
-        }
-    }
-
-    private void updateCurrentTimeInSeconds() {
-        currentTimeInSeconds = TimeUnit.MILLISECONDS.toSeconds(clock.milliTime());
     }
 
     public int size() {
@@ -236,6 +154,10 @@ class ExponentiallyDecayingReservoir {
         } finally {
             unlockForRegularUsage();
         }
+    }
+
+    private void updateTimeInSeconds() {
+        currentTimeInSeconds = System.currentTimeMillis() / 1000;
     }
 
     long currentTimeInSeconds() {

--- a/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
@@ -23,7 +23,6 @@ import java.util.concurrent.ThreadLocalRandom;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
-import java.util.logging.Logger;
 
 import static java.lang.Math.exp;
 import static java.lang.Math.min;

--- a/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
@@ -154,7 +154,7 @@ class ExponentiallyDecayingReservoir {
     }
 
     private void updateTimeInSeconds() {
-        currentTimeInSeconds = System.currentTimeMillis() / 1000;
+        currentTimeInSeconds = TimeUnit.MILLISECONDS.toSeconds(clock.milliTime());
     }
 
     long currentTimeInSeconds() {

--- a/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
@@ -23,8 +23,8 @@ import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ThreadLocalRandom;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 import java.util.logging.Level;
 import java.util.logging.Logger;
@@ -49,6 +49,11 @@ import static java.lang.Math.min;
  * @see <a href="http://dimacs.rutgers.edu/~graham/pubs/papers/fwddecay.pdf">
  * Cormode et al. Forward Decay: A Practical Time Decay Model for Streaming Systems. ICDE '09:
  * Proceedings of the 2009 IEEE International Conference on Data Engineering (2009)</a>
+ *
+ * To avoid calculating the current time in seconds during every update, use an executor that runs just a few times each second
+ * to update the current-time-in-seconds value for each instance. Each instance might have a separate {@code Clock} for obtaining
+ * the current time, so we cannot share a single static value for the current time across all instances. So each instance
+ * registers its own {@code Runnable} which updates its own value, and the single executor invokes all of them when it runs.
  */
 class ExponentiallyDecayingReservoir {
     private static final int DEFAULT_SIZE = 1028;
@@ -60,11 +65,20 @@ class ExponentiallyDecayingReservoir {
      */
     private static final long CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS = 250;
 
-    private static final List<ExponentiallyDecayingReservoir> reservoirs = new ArrayList<>();
+    private static final List<ExponentiallyDecayingReservoir> RESERVOIRS = new ArrayList<>();
 
-    private static final AtomicBoolean EXECUTOR_UPDATER_CONFIGURED = new AtomicBoolean();
     private static final Logger LOGGER = Logger.getLogger(ExponentiallyDecayingReservoir.class.getName());
     private static volatile ScheduledExecutorService currentTimeUpdaterExecutorService;
+
+    private enum CurrentTimeExecutorServiceState {
+        DORMANT, // never started
+        STARTED, // started and still running
+        STOPPED  // stopped
+        }
+
+    private static final AtomicReference<CurrentTimeExecutorServiceState> CURRENT_TIME_EXECUTOR_SERVICE_STATE =
+            new AtomicReference<>(CurrentTimeExecutorServiceState.DORMANT);
+
     private final ConcurrentSkipListMap<Double, WeightedSnapshot.WeightedSample> values;
     private final ReentrantReadWriteLock lock;
     private final double alpha;
@@ -96,49 +110,73 @@ class ExponentiallyDecayingReservoir {
         this.alpha = alpha;
         this.size = size;
         this.count = new AtomicLong(0);
-        reservoirs.add(this);
+        RESERVOIRS.add(this);
         updateCurrentTimeInSeconds();
         this.startTime = currentTimeInSeconds;
         this.nextScaleTime = new AtomicLong(clock.nanoTick() + RESCALE_THRESHOLD);
     }
 
     static void init() {
-        if (EXECUTOR_UPDATER_CONFIGURED.compareAndSet(false, true)) {
-            ScheduledExecutorService result = Executors.newSingleThreadScheduledExecutor();
-            result.scheduleAtFixedRate(ExponentiallyDecayingReservoir::updateCurrentTimeInSecondsForAllReservoirs,
-                                       CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS,
-                                       CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS, TimeUnit.MILLISECONDS);
-            currentTimeUpdaterExecutorService = result;
-        }
+        CURRENT_TIME_EXECUTOR_SERVICE_STATE.getAndUpdate(ExponentiallyDecayingReservoir::startCurrentTimeExecutor);
     }
 
     // we only expect a single instance of MetricSupport registered
     // with webserver, so we can stop the executor when the associated webserver stops
     static void onServerShutdown() {
-        ScheduledExecutorService service = currentTimeUpdaterExecutorService;
-        if (EXECUTOR_UPDATER_CONFIGURED.compareAndSet(true, false)) {
-            try {
-                boolean stoppedNormally =
-                        service.awaitTermination(CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS * 10,
-                                                 TimeUnit.MILLISECONDS);
-                if (!stoppedNormally) {
-                    LOGGER.log(Level.WARNING, "Shutdown of current time updater timed out; continuing");
-                }
-            } catch (InterruptedException e) {
-                LOGGER.log(Level.WARNING, "InterruptedException caught while stopping the current time updater; continuing");
-            }
+        CURRENT_TIME_EXECUTOR_SERVICE_STATE.getAndUpdate(ExponentiallyDecayingReservoir::stopCurrentTimeExecutor);
+    }
+
+    private static CurrentTimeExecutorServiceState startCurrentTimeExecutor(CurrentTimeExecutorServiceState currentState) {
+        CurrentTimeExecutorServiceState newState = currentState;
+        if (currentState == CurrentTimeExecutorServiceState.DORMANT) {
+            LOGGER.log(Level.FINE, () -> String.format("Starting %s current time executor with interval %d ms",
+                    ExponentiallyDecayingReservoir.class.getSimpleName(), CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS));
+            newState = CurrentTimeExecutorServiceState.STARTED;
+            ScheduledExecutorService result = Executors.newSingleThreadScheduledExecutor();
+            result.scheduleAtFixedRate(ExponentiallyDecayingReservoir::updateCurrentTimeInSecondsForAllReservoirs,
+                    CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS,
+                    CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS, TimeUnit.MILLISECONDS);
+            currentTimeUpdaterExecutorService = result;
+        } else {
+            LOGGER.log(Level.WARNING, String.format(
+                    "Attempt to start current time executor; expected state %s but found %s; ignored",
+                    CurrentTimeExecutorServiceState.DORMANT,
+                    currentState), new IllegalStateException());
         }
+        return newState;
+    }
+
+    private static CurrentTimeExecutorServiceState stopCurrentTimeExecutor(CurrentTimeExecutorServiceState currentState) {
+        switch (currentState) {
+            case STARTED:
+                shutdownCurrentTimeExecutor();
+                break;
+
+            case STOPPED:
+            case DORMANT:
+                break;
+
+            default:
+                LOGGER.log(Level.WARNING, "Attempt to stop current time executor found unexpected state " + currentState);
+                shutdownCurrentTimeExecutor();
+        }
+        return CurrentTimeExecutorServiceState.STOPPED;
+    }
+
+    private static void shutdownCurrentTimeExecutor() {
+        LOGGER.log(Level.FINE, String.format("Shutting down %s current time executor",
+                ExponentiallyDecayingReservoir.class.getSimpleName()));
+        currentTimeUpdaterExecutorService.shutdownNow();
     }
 
     private static void updateCurrentTimeInSecondsForAllReservoirs() {
-        reservoirs.forEach(ExponentiallyDecayingReservoir::updateCurrentTimeInSeconds);
+        for (ExponentiallyDecayingReservoir reservoir : RESERVOIRS) {
+            reservoir.updateCurrentTimeInSeconds();
+        }
     }
 
     private void updateCurrentTimeInSeconds() {
         currentTimeInSeconds = TimeUnit.MILLISECONDS.toSeconds(clock.milliTime());
-        if (LOGGER.isLoggable(Level.FINER)) {
-            LOGGER.log(Level.FINER, "Updating current time in seconds for " + hashCode() + " to " + currentTimeInSeconds);
-        }
     }
 
     public int size() {
@@ -202,7 +240,7 @@ class ExponentiallyDecayingReservoir {
     long currentTimeInSeconds() {
         return currentTimeInSeconds;
     }
-    
+
     private double weight(long t) {
         return exp(alpha * t);
     }

--- a/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
@@ -23,6 +23,7 @@ import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ThreadLocalRandom;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 import java.util.logging.Level;
@@ -61,42 +62,9 @@ class ExponentiallyDecayingReservoir {
 
     private static final List<Runnable> CURRENT_TIME_IN_SECONDS_UPDATERS = new ArrayList<>();
 
-    private static final ScheduledExecutorService CURRENT_TIME_UPDATER_EXECUTOR_SERVICE = initCurrentTimeUpdater();
-
+    private static final AtomicBoolean EXECUTOR_UPDATER_CONFIGURED = new AtomicBoolean();
     private static final Logger LOGGER = Logger.getLogger(ExponentiallyDecayingReservoir.class.getName());
-
-    private volatile long currentTimeInSeconds;
-
-    private static ScheduledExecutorService initCurrentTimeUpdater() {
-        ScheduledExecutorService result = Executors.newSingleThreadScheduledExecutor();
-        result.scheduleAtFixedRate(ExponentiallyDecayingReservoir::updateCurrentTimeInSecondsForAllReservoirs,
-                CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS,
-                CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS, TimeUnit.MILLISECONDS);
-        return result;
-    }
-
-    static void onServerShutdown() {
-        CURRENT_TIME_UPDATER_EXECUTOR_SERVICE.shutdown();
-        try {
-            boolean stoppedNormally =
-                    CURRENT_TIME_UPDATER_EXECUTOR_SERVICE.awaitTermination(CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS * 10,
-                            TimeUnit.MILLISECONDS);
-            if (!stoppedNormally) {
-                LOGGER.log(Level.WARNING, "Shutdown of current time updater timed out; continuing");
-            }
-        } catch (InterruptedException e) {
-            LOGGER.log(Level.WARNING, "InterruptedException caught while stopping the current time updater; continuing");
-        }
-    }
-
-    private static void updateCurrentTimeInSecondsForAllReservoirs() {
-        CURRENT_TIME_IN_SECONDS_UPDATERS.forEach(Runnable::run);
-    }
-
-    private long computeCurrentTimeInSeconds() {
-        return TimeUnit.MILLISECONDS.toSeconds(clock.milliTime());
-    }
-
+    private static volatile ScheduledExecutorService currentTimeUpdaterExecutorService;
     private final ConcurrentSkipListMap<Double, WeightedSnapshot.WeightedSample> values;
     private final ReentrantReadWriteLock lock;
     private final double alpha;
@@ -104,8 +72,8 @@ class ExponentiallyDecayingReservoir {
     private final AtomicLong count;
     private final Clock clock;
     private final AtomicLong nextScaleTime;
+    private volatile long currentTimeInSeconds;
     private volatile long startTime;
-
     /**
      * Creates a new {@link ExponentiallyDecayingReservoir} of 1028 elements, which offers a 99.9%
      * confidence level with a 5% margin of error assuming a normal distribution, and an alpha
@@ -114,7 +82,6 @@ class ExponentiallyDecayingReservoir {
     ExponentiallyDecayingReservoir(Clock clock) {
         this(DEFAULT_SIZE, DEFAULT_ALPHA, clock);
     }
-
     /**
      * Creates a new {@link ExponentiallyDecayingReservoir}.
      *
@@ -133,6 +100,42 @@ class ExponentiallyDecayingReservoir {
         currentTimeInSeconds = computeCurrentTimeInSeconds();
         this.startTime = currentTimeInSeconds;
         this.nextScaleTime = new AtomicLong(clock.nanoTick() + RESCALE_THRESHOLD);
+    }
+
+    static void init() {
+        if (EXECUTOR_UPDATER_CONFIGURED.compareAndSet(false, true)) {
+            ScheduledExecutorService result = Executors.newSingleThreadScheduledExecutor();
+            result.scheduleAtFixedRate(ExponentiallyDecayingReservoir::updateCurrentTimeInSecondsForAllReservoirs,
+                                       CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS,
+                                       CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS, TimeUnit.MILLISECONDS);
+            currentTimeUpdaterExecutorService = result;
+        }
+    }
+
+    // we only expect a single instance of MetricSupport registered
+    // with webserver, so we can stop the executor when the associated webserver stops
+    static void onServerShutdown() {
+        ScheduledExecutorService service = currentTimeUpdaterExecutorService;
+        if (EXECUTOR_UPDATER_CONFIGURED.compareAndSet(true, false)) {
+            try {
+                boolean stoppedNormally =
+                        service.awaitTermination(CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS * 10,
+                                                 TimeUnit.MILLISECONDS);
+                if (!stoppedNormally) {
+                    LOGGER.log(Level.WARNING, "Shutdown of current time updater timed out; continuing");
+                }
+            } catch (InterruptedException e) {
+                LOGGER.log(Level.WARNING, "InterruptedException caught while stopping the current time updater; continuing");
+            }
+        }
+    }
+
+    private static void updateCurrentTimeInSecondsForAllReservoirs() {
+        CURRENT_TIME_IN_SECONDS_UPDATERS.forEach(Runnable::run);
+    }
+
+    private long computeCurrentTimeInSeconds() {
+        return TimeUnit.MILLISECONDS.toSeconds(clock.milliTime());
     }
 
     public int size() {

--- a/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
@@ -24,7 +24,6 @@ import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ThreadLocalRandom;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicLong;
-import java.util.concurrent.atomic.AtomicReference;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 import java.util.logging.Level;
 import java.util.logging.Logger;
@@ -76,8 +75,7 @@ class ExponentiallyDecayingReservoir {
         STOPPED  // stopped
         }
 
-    private static final AtomicReference<CurrentTimeExecutorServiceState> CURRENT_TIME_EXECUTOR_SERVICE_STATE =
-            new AtomicReference<>(CurrentTimeExecutorServiceState.DORMANT);
+    private static CurrentTimeExecutorServiceState currentTimeExecutorServiceState = CurrentTimeExecutorServiceState.DORMANT;
 
     private final ConcurrentSkipListMap<Double, WeightedSnapshot.WeightedSample> values;
     private final ReentrantReadWriteLock lock;
@@ -116,38 +114,41 @@ class ExponentiallyDecayingReservoir {
         this.nextScaleTime = new AtomicLong(clock.nanoTick() + RESCALE_THRESHOLD);
     }
 
-    static void init() {
-        CURRENT_TIME_EXECUTOR_SERVICE_STATE.getAndUpdate(ExponentiallyDecayingReservoir::startCurrentTimeExecutor);
+    static synchronized void init() {
+        if (currentTimeExecutorServiceState == CurrentTimeExecutorServiceState.DORMANT) {
+            currentTimeExecutorServiceState = CurrentTimeExecutorServiceState.STARTED;
+            currentTimeUpdaterExecutorService = startCurrentTimeExecutor();
+        }
     }
 
     // we only expect a single instance of MetricSupport registered
     // with webserver, so we can stop the executor when the associated webserver stops
-    static void onServerShutdown() {
-        CURRENT_TIME_EXECUTOR_SERVICE_STATE.getAndUpdate(ExponentiallyDecayingReservoir::stopCurrentTimeExecutor);
-    }
-
-    private static CurrentTimeExecutorServiceState startCurrentTimeExecutor(CurrentTimeExecutorServiceState currentState) {
-        CurrentTimeExecutorServiceState newState = currentState;
-        if (currentState == CurrentTimeExecutorServiceState.DORMANT) {
-            LOGGER.log(Level.FINE, () -> String.format("Starting %s current time executor with interval %d ms",
-                    ExponentiallyDecayingReservoir.class.getSimpleName(), CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS));
-            newState = CurrentTimeExecutorServiceState.STARTED;
-            ScheduledExecutorService result = Executors.newSingleThreadScheduledExecutor();
-            result.scheduleAtFixedRate(ExponentiallyDecayingReservoir::updateCurrentTimeInSecondsForAllReservoirs,
-                    CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS,
-                    CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS, TimeUnit.MILLISECONDS);
-            currentTimeUpdaterExecutorService = result;
+    static synchronized void onServerShutdown() {
+        if (currentTimeExecutorServiceState == CurrentTimeExecutorServiceState.STARTED) {
+            stopCurrentTimeExecutor();
         } else {
             LOGGER.log(Level.WARNING, String.format(
                     "Attempt to start current time executor; expected state %s but found %s; ignored",
                     CurrentTimeExecutorServiceState.DORMANT,
-                    currentState), new IllegalStateException());
+                    currentTimeExecutorServiceState),
+                new IllegalStateException());
         }
-        return newState;
+        currentTimeExecutorServiceState = CurrentTimeExecutorServiceState.STOPPED;
     }
 
-    private static CurrentTimeExecutorServiceState stopCurrentTimeExecutor(CurrentTimeExecutorServiceState currentState) {
-        switch (currentState) {
+    private static ScheduledExecutorService startCurrentTimeExecutor() {
+        LOGGER.log(Level.FINE, () -> String.format("Starting %s current time executor with interval %d ms",
+                ExponentiallyDecayingReservoir.class.getSimpleName(), CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS));
+        ScheduledExecutorService result = Executors.newSingleThreadScheduledExecutor();
+        result.scheduleAtFixedRate(ExponentiallyDecayingReservoir::updateCurrentTimeInSecondsForAllReservoirs,
+                CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS,
+                CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS,
+                TimeUnit.MILLISECONDS);
+        return result;
+    }
+
+    private static void stopCurrentTimeExecutor() {
+        switch (currentTimeExecutorServiceState) {
             case STARTED:
                 shutdownCurrentTimeExecutor();
                 break;
@@ -157,10 +158,10 @@ class ExponentiallyDecayingReservoir {
                 break;
 
             default:
-                LOGGER.log(Level.WARNING, "Attempt to stop current time executor found unexpected state " + currentState);
+                LOGGER.log(Level.WARNING, "Attempt to stop current time executor found unexpected state "
+                        + currentTimeExecutorServiceState);
                 shutdownCurrentTimeExecutor();
         }
-        return CurrentTimeExecutorServiceState.STOPPED;
     }
 
     private static void shutdownCurrentTimeExecutor() {

--- a/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
@@ -60,7 +60,7 @@ class ExponentiallyDecayingReservoir {
      */
     private static final long CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL_MS = 250;
 
-    private static final List<Runnable> CURRENT_TIME_IN_SECONDS_UPDATERS = new ArrayList<>();
+    private static final List<ExponentiallyDecayingReservoir> reservoirs = new ArrayList<>();
 
     private static final AtomicBoolean EXECUTOR_UPDATER_CONFIGURED = new AtomicBoolean();
     private static final Logger LOGGER = Logger.getLogger(ExponentiallyDecayingReservoir.class.getName());
@@ -96,8 +96,8 @@ class ExponentiallyDecayingReservoir {
         this.alpha = alpha;
         this.size = size;
         this.count = new AtomicLong(0);
-        CURRENT_TIME_IN_SECONDS_UPDATERS.add(this::computeCurrentTimeInSeconds);
-        currentTimeInSeconds = computeCurrentTimeInSeconds();
+        reservoirs.add(this);
+        updateCurrentTimeInSeconds();
         this.startTime = currentTimeInSeconds;
         this.nextScaleTime = new AtomicLong(clock.nanoTick() + RESCALE_THRESHOLD);
     }
@@ -131,11 +131,14 @@ class ExponentiallyDecayingReservoir {
     }
 
     private static void updateCurrentTimeInSecondsForAllReservoirs() {
-        CURRENT_TIME_IN_SECONDS_UPDATERS.forEach(Runnable::run);
+        reservoirs.forEach(ExponentiallyDecayingReservoir::updateCurrentTimeInSeconds);
     }
 
-    private long computeCurrentTimeInSeconds() {
-        return TimeUnit.MILLISECONDS.toSeconds(clock.milliTime());
+    private void updateCurrentTimeInSeconds() {
+        currentTimeInSeconds = TimeUnit.MILLISECONDS.toSeconds(clock.milliTime());
+        if (LOGGER.isLoggable(Level.FINER)) {
+            LOGGER.log(Level.FINER, "Updating current time in seconds for " + hashCode() + " to " + currentTimeInSeconds);
+        }
     }
 
     public int size() {

--- a/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
@@ -57,8 +57,6 @@ class ExponentiallyDecayingReservoir {
     private static final double DEFAULT_ALPHA = 0.015;
     private static final long RESCALE_THRESHOLD = TimeUnit.HOURS.toNanos(1);
 
-    private static final Logger LOGGER = Logger.getLogger(ExponentiallyDecayingReservoir.class.getName());
-
     private static final Duration CURRENT_TIME_IN_SECONDS_UPDATE_INTERVAL = Duration.ofMillis(250);
 
     private final ConcurrentSkipListMap<Double, WeightedSnapshot.WeightedSample> values;

--- a/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/ExponentiallyDecayingReservoir.java
@@ -196,6 +196,10 @@ class ExponentiallyDecayingReservoir {
         }
     }
 
+    long currentTimeInSeconds() {
+        return currentTimeInSeconds;
+    }
+    
     private double weight(long t) {
         return exp(alpha * t);
     }

--- a/metrics/metrics/src/main/java/io/helidon/metrics/MetricsSupport.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/MetricsSupport.java
@@ -427,6 +427,8 @@ public final class MetricsSupport extends HelidonRestServiceSupport {
      */
     @Override
     public void update(Routing.Rules rules) {
+        ExponentiallyDecayingReservoir.init();
+
         configureEndpoint(rules, rules);
     }
 

--- a/metrics/metrics/src/main/java/io/helidon/metrics/MetricsSupport.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/MetricsSupport.java
@@ -427,14 +427,14 @@ public final class MetricsSupport extends HelidonRestServiceSupport {
      */
     @Override
     public void update(Routing.Rules rules) {
-        ExponentiallyDecayingReservoir.init();
+        PeriodicExecutor.start();
 
         configureEndpoint(rules, rules);
     }
 
     @Override
     protected void onShutdown() {
-        ExponentiallyDecayingReservoir.onServerShutdown();
+        PeriodicExecutor.stop();
     }
 
     private static KeyPerformanceIndicatorSupport.Context kpiContext(ServerRequest request) {

--- a/metrics/metrics/src/main/java/io/helidon/metrics/PeriodicExecutor.java
+++ b/metrics/metrics/src/main/java/io/helidon/metrics/PeriodicExecutor.java
@@ -1,0 +1,144 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.metrics;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Manages a single-thread executor which will invoke callbacks to enrollees at a frequency they request.
+ * <p>
+ * Some enrollments might arrive before the manager is started. We save those and act on them once the
+ * manager starts. This makes sure the executor's thread starts only at native image runtime.
+ * </p>
+ * <p>
+ * In production use, starting and stopping the executor and even enrolling callbacks are not performance-critical operations,
+ * so simple synchronization on methods which access shared data is clear and sufficient.
+ * </p>
+ */
+class PeriodicExecutor {
+
+    static final PeriodicExecutor INSTANCE = create();
+
+    static PeriodicExecutor create() {
+        return new PeriodicExecutor();
+    }
+
+    private static final Logger LOGGER = Logger.getLogger(PeriodicExecutor.class.getName());
+
+    enum State {
+        DORMANT, // never started
+        STARTED, // started and still running
+        STOPPED  // stopped
+    }
+
+    private static class Enrollment {
+        private final Runnable runnable;
+        private final Duration interval;
+
+        Enrollment(Runnable runnable, Duration interval) {
+            this.runnable = runnable;
+            this.interval = interval;
+        }
+    }
+
+    private State state = State.DORMANT;
+
+    private ScheduledExecutorService currentTimeUpdaterExecutorService;
+
+    private final Collection<Enrollment> deferredEnrollments = new ArrayList<>();
+
+    private PeriodicExecutor() {
+    }
+
+    static void enroll(Runnable runnable, Duration interval) {
+        INSTANCE.enrollRunner(runnable, interval);
+    }
+
+    static void start() {
+        INSTANCE.startExecutor();
+    }
+
+    static void stop() {
+        INSTANCE.stopExecutor();
+    }
+
+    static State state() {
+        return INSTANCE.executorState();
+    }
+
+    synchronized void enrollRunner(Runnable runnable, Duration interval) {
+
+        if (state == State.STARTED) {
+            currentTimeUpdaterExecutorService.scheduleAtFixedRate(
+                    runnable,
+                    interval.toMillis(),
+                    interval.toMillis(),
+                    TimeUnit.MILLISECONDS);
+        } else if (state == State.DORMANT) {
+            deferredEnrollments.add(new Enrollment(runnable, interval));
+        } else {
+            LOGGER.log(Level.WARNING,
+                    "Attempt to enroll when in unexpected state " + state + "; ignored",
+                    new IllegalStateException());
+        }
+    }
+
+    synchronized void startExecutor() {
+        if (state == State.DORMANT) {
+            LOGGER.log(Level.FINE, "Starting up with " + deferredEnrollments.size() + " deferred enrollments");
+            state = State.STARTED;
+            currentTimeUpdaterExecutorService = Executors.newSingleThreadScheduledExecutor();
+            for (Enrollment deferredEnrollment : deferredEnrollments) {
+                currentTimeUpdaterExecutorService.scheduleAtFixedRate(
+                        deferredEnrollment.runnable,
+                        deferredEnrollment.interval.toMillis(),
+                        deferredEnrollment.interval.toMillis(),
+                        TimeUnit.MILLISECONDS);
+            }
+            deferredEnrollments.clear();
+        } else {
+            LOGGER.log(Level.WARNING, String.format("Attempt to start; the expected state is %s but found %s; ignored",
+                    State.DORMANT,
+                    state),
+                    new IllegalStateException());
+        }
+    }
+
+    synchronized void stopExecutor() {
+        if (state == State.STARTED) {
+            LOGGER.log(Level.FINE, "Shutting down");
+            currentTimeUpdaterExecutorService.shutdownNow();
+        } else {
+            LOGGER.log(Level.WARNING, String.format(
+                    "Attempt to stop; the expected state is %s but found %s; ignored",
+                    State.DORMANT,
+                    state),
+                    new IllegalStateException());
+        }
+        state = State.STOPPED;
+    }
+
+    State executorState() {
+        return state;
+    }
+}

--- a/metrics/metrics/src/main/resources/META-INF/native-image/io.helidon.metrics/native-image.properties
+++ b/metrics/metrics/src/main/resources/META-INF/native-image/io.helidon.metrics/native-image.properties
@@ -14,4 +14,4 @@
 # limitations under the License.
 #
 
-Args=--initialize-at-run-time=io.helidon.metrics.ExponentiallyDecayingReservoir
+Args=--initialize-at-build-time=io.helidon.metrics

--- a/metrics/metrics/src/test/java/io/helidon/metrics/TestExponentiallyDecayingReservoir.java
+++ b/metrics/metrics/src/test/java/io/helidon/metrics/TestExponentiallyDecayingReservoir.java
@@ -27,12 +27,12 @@ class TestExponentiallyDecayingReservoir {
 
     @BeforeAll
     static void startExecutor() {
-        ExponentiallyDecayingReservoir.init();
+        PeriodicExecutor.start();
     }
 
     @AfterAll
     static void stopExecutor() {
-        ExponentiallyDecayingReservoir.onServerShutdown();
+        PeriodicExecutor.stop();
     }
 
     @Test

--- a/metrics/metrics/src/test/java/io/helidon/metrics/TestExponentiallyDecayingReservoir.java
+++ b/metrics/metrics/src/test/java/io/helidon/metrics/TestExponentiallyDecayingReservoir.java
@@ -1,0 +1,47 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.metrics;
+
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.is;
+
+class TestExponentiallyDecayingReservoir {
+
+    @BeforeAll
+    static void startExecutor() {
+        ExponentiallyDecayingReservoir.init();
+    }
+
+    @AfterAll
+    static void stopExecutor() {
+        ExponentiallyDecayingReservoir.onServerShutdown();
+    }
+
+    @Test
+    void checkCurrentTimeInSeconds() throws InterruptedException, NoSuchFieldException, IllegalAccessException {
+        ExponentiallyDecayingReservoir edr = new ExponentiallyDecayingReservoir(Clock.system());
+        long startTime = edr.currentTimeInSeconds();
+        Thread.sleep(1100);
+        long currentTimeInSeconds = edr.currentTimeInSeconds();
+        assertThat("Difference in current time across a short delay", currentTimeInSeconds - startTime,
+                is(greaterThan(0L)));
+    }
+}

--- a/metrics/metrics/src/test/java/io/helidon/metrics/TestExponentiallyDecayingReservoir.java
+++ b/metrics/metrics/src/test/java/io/helidon/metrics/TestExponentiallyDecayingReservoir.java
@@ -36,12 +36,11 @@ class TestExponentiallyDecayingReservoir {
     }
 
     @Test
-    void checkCurrentTimeInSeconds() throws InterruptedException, NoSuchFieldException, IllegalAccessException {
+    void checkCurrentTimeInSeconds() throws InterruptedException {
         ExponentiallyDecayingReservoir edr = new ExponentiallyDecayingReservoir(Clock.system());
         long startTime = edr.currentTimeInSeconds();
         Thread.sleep(1100);
-        long currentTimeInSeconds = edr.currentTimeInSeconds();
-        assertThat("Difference in current time across a short delay", currentTimeInSeconds - startTime,
+        assertThat("Difference in current time across a short delay", edr.currentTimeInSeconds() - startTime,
                 is(greaterThan(0L)));
     }
 }

--- a/metrics/metrics/src/test/java/io/helidon/metrics/TestPeriodicExecutor.java
+++ b/metrics/metrics/src/test/java/io/helidon/metrics/TestPeriodicExecutor.java
@@ -1,0 +1,110 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.metrics;
+
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.is;
+
+class TestPeriodicExecutor {
+
+    private static final int SLEEP_TIME_MS = 1500;
+
+    private static final int FAST_INTERVAL = 250;
+    private static final int SLOW_INTERVAL = 400;
+
+    private static final double SLOWDOWN_FACTOR = 0.80; // for slow pipelines!
+
+    private static final double MIN_FAST_COUNT = 1500 / FAST_INTERVAL * SLOWDOWN_FACTOR;
+    private static final double MIN_SLOW_COUNT = 1500 / SLOW_INTERVAL * SLOWDOWN_FACTOR;
+
+    @Test
+    void testWithNoDeferrals() throws InterruptedException {
+
+        PeriodicExecutor exec = PeriodicExecutor.create();
+        try {
+            exec.startExecutor();
+            AtomicInteger countA = new AtomicInteger();
+            AtomicInteger countB = new AtomicInteger();
+
+            exec.enrollRunner(() -> countA.incrementAndGet(), Duration.ofMillis(FAST_INTERVAL));
+            exec.enrollRunner(() -> countB.incrementAndGet(), Duration.ofMillis(SLOW_INTERVAL));
+
+            Thread.sleep(SLEEP_TIME_MS);
+
+            assertThat("CountA", (double) countA.get(), is(greaterThan(MIN_FAST_COUNT)));
+            assertThat("CountB", (double) countB.get(), is(greaterThan(MIN_SLOW_COUNT)));
+        } finally {
+            if (exec.executorState() == PeriodicExecutor.State.STARTED) {
+                exec.stopExecutor();
+            }
+        }
+    }
+
+    @Test
+    void testWithDeferredEnrollments() throws InterruptedException {
+        PeriodicExecutor exec = PeriodicExecutor.create();
+        try {
+            AtomicInteger countA = new AtomicInteger();
+            AtomicInteger countB = new AtomicInteger();
+
+            exec.enrollRunner(() -> countA.incrementAndGet(), Duration.ofMillis(FAST_INTERVAL));
+
+            exec.startExecutor();
+
+            exec.enrollRunner(() -> countB.incrementAndGet(), Duration.ofMillis(SLOW_INTERVAL));
+
+            Thread.sleep(SLEEP_TIME_MS);
+
+            assertThat("CountA", (double) countA.get(), is(greaterThan(MIN_FAST_COUNT)));
+            assertThat("CountB", (double) countB.get(), is(greaterThan(MIN_SLOW_COUNT)));
+        } finally {
+            if (exec.executorState() == PeriodicExecutor.State.STARTED) {
+                exec.stopExecutor();
+            }
+        }
+    }
+
+    @Test
+    void testWithLateEnrollment() throws InterruptedException {
+        PeriodicExecutor exec = PeriodicExecutor.create();
+        try {
+            AtomicInteger countA = new AtomicInteger();
+            AtomicInteger countB = new AtomicInteger();
+
+            exec.enrollRunner(() -> countA.incrementAndGet(), Duration.ofMillis(FAST_INTERVAL));
+
+            exec.startExecutor();
+            Thread.sleep(SLEEP_TIME_MS);
+
+            exec.stopExecutor();
+
+            exec.enrollRunner(() -> countB.incrementAndGet(), Duration.ofMillis(SLOW_INTERVAL));
+
+            assertThat("CountA", (double) countA.get(), is(greaterThan(MIN_FAST_COUNT))); // should be 8
+            assertThat("CountB", (double) countB.get(), is(0.0));
+        } finally {
+            if (exec.executorState() == PeriodicExecutor.State.STARTED) {
+                exec.stopExecutor();
+            }
+        }
+    }
+}

